File: go/adxl355/transport_contract_test.go
package adxl355

import (
	"errors"
	"testing"
)

type contractTransport struct {
	regs         [128]byte
	shortReg     int
	shortLength  int
	failReadReg  int
	failWriteReg int
}

func newContractTransport() *contractTransport {
	transport := &contractTransport{shortReg: -1, failReadReg: -1, failWriteReg: -1}
	transport.regs[RegDEVID_AD] = DEVID_AD_VALUE
	transport.regs[RegDEVID_MST] = DEVID_MST_VALUE
	transport.regs[RegPARTID] = PARTID_VALUE
	transport.regs[RegRANGE] = byte(Range2G)
	transport.regs[RegPOWER_CTL] = byte(PowerStandby)
	return transport
}

func (t *contractTransport) ReadRegister(reg byte, length int) ([]byte, error) {
	if t.failReadReg == int(reg) {
		return nil, errors.New("native read failure")
	}
	returned := length
	if t.shortReg == int(reg) {
		returned = t.shortLength
	}
	return t.regs[int(reg) : int(reg)+returned], nil
}

func (t *contractTransport) WriteRegister(reg byte, data []byte) error {
	if t.failWriteReg == int(reg) {
		return errors.New("native write failure")
	}
	copy(t.regs[reg:], data)
	return nil
}

func (t *contractTransport) DelayMs(_ uint32) {}

func probedContractDevice(t *testing.T, transport *contractTransport) *Device {
	t.Helper()
	device := New(transport)
	if _, err := device.Probe(); err != nil {
		t.Fatalf("Probe failed: %v", err)
	}
	return device
}

func TestTransportContractSingleRegisterExactLength(t *testing.T) {
	for _, returned := range []int{0, 2} {
		transport := newContractTransport()
		transport.shortReg = RegDEVID_AD
		transport.shortLength = returned
		if _, err := New(transport).Probe(); err != ErrBus {
			t.Fatalf("returned=%d error=%v, want ErrBus", returned, err)
		}
	}
}

func TestTransportContractTemperatureExactLength(t *testing.T) {
	for _, returned := range []int{0, 1} {
		transport := newContractTransport()
		device := probedContractDevice(t, transport)
		transport.shortReg = RegTEMP2
		transport.shortLength = returned
		if _, err := device.ReadTemperatureRaw(); err != ErrBus {
			t.Fatalf("returned=%d error=%v, want ErrBus", returned, err)
		}
	}
}

func TestTransportContractXYZExactLength(t *testing.T) {
	for _, returned := range []int{0, 8} {
		transport := newContractTransport()
		device := probedContractDevice(t, transport)
		transport.shortReg = RegXDATA3
		transport.shortLength = returned
		if _, err := device.ReadRaw(); err != ErrBus {
			t.Fatalf("returned=%d error=%v, want ErrBus", returned, err)
		}
	}
}

func TestTransportErrorsNormalizeToBus(t *testing.T) {
	transport := newContractTransport()
	transport.failReadReg = RegDEVID_AD
	if _, err := New(transport).Probe(); err != ErrBus {
		t.Fatalf("read error=%v, want ErrBus", err)
	}

	transport = newContractTransport()
	device := probedContractDevice(t, transport)
	transport.failWriteReg = RegRANGE
	if err := device.SetRange(Range4G); err != ErrBus {
		t.Fatalf("write error=%v, want ErrBus", err)
	}
}

File: node/test/transportContract.test.ts
import { describe, expect, it } from "vitest";
import { ADXL355 } from "../src/device.js";
import { BusError } from "../src/errors.js";
import { Range, Reg } from "../src/registers.js";
import { Transport } from "../src/transport.js";

class ContractTransport implements Transport {
  private readonly regs = new Uint8Array(128);
  shortReg: number | undefined;
  shortLength = 0;
  failReadReg: number | undefined;
  failWriteReg: number | undefined;
  failDelay = false;

  constructor() {
    this.regs[Reg.DEVID_AD] = 0xad;
    this.regs[Reg.DEVID_MST] = 0x1d;
    this.regs[Reg.PARTID] = 0xed;
    this.regs[Reg.RANGE] = Range.G2;
    this.regs[Reg.POWER_CTL] = 1;
  }

  async readRegister(reg: number, length: number): Promise<Uint8Array> {
    if (this.failReadReg === reg) {
      throw new TypeError("native read failure");
    }
    const returned = this.shortReg === reg ? this.shortLength : length;
    return this.regs.slice(reg, reg + returned);
  }

  async writeRegister(reg: number, data: Uint8Array): Promise<void> {
    if (this.failWriteReg === reg) {
      throw new TypeError("native write failure");
    }
    this.regs.set(data, reg);
  }

  async delayMs(_ms: number): Promise<void> {
    if (this.failDelay) {
      throw new TypeError("native delay failure");
    }
  }
}

async function probedDevice(transport = new ContractTransport()): Promise<ADXL355> {
  const device = new ADXL355(transport);
  await device.probe();
  return device;
}

describe("shared transport contract", () => {
  for (const [name, returned] of [["TR-1-ZERO", 0], ["TR-1-OVERLONG", 2]] as const) {
    it(name, async () => {
      const transport = new ContractTransport();
      transport.shortReg = Reg.DEVID_AD;
      transport.shortLength = returned;
      await expect(new ADXL355(transport).probe()).rejects.toBeInstanceOf(BusError);
    });
  }

  for (const [name, returned] of [["TR-2-ZERO", 0], ["TR-2-TRUNCATED", 1]] as const) {
    it(name, async () => {
      const transport = new ContractTransport();
      const device = await probedDevice(transport);
      transport.shortReg = Reg.TEMP2;
      transport.shortLength = returned;
      await expect(device.readTemperatureRaw()).rejects.toBeInstanceOf(BusError);
    });
  }

  for (const [name, returned] of [["TR-9-ZERO", 0], ["TR-9-TRUNCATED", 8]] as const) {
    it(name, async () => {
      const transport = new ContractTransport();
      const device = await probedDevice(transport);
      transport.shortReg = Reg.XDATA3;
      transport.shortLength = returned;
      await expect(device.readRaw()).rejects.toBeInstanceOf(BusError);
    });
  }

  it("normalizes native read failures", async () => {
    const transport = new ContractTransport();
    transport.failReadReg = Reg.DEVID_AD;
    await expect(new ADXL355(transport).probe()).rejects.toBeInstanceOf(BusError);
  });

  it("normalizes native write failures", async () => {
    const transport = new ContractTransport();
    const device = await probedDevice(transport);
    transport.failWriteReg = Reg.RANGE;
    await expect(device.setRange(Range.G4)).rejects.toBeInstanceOf(BusError);
  });

  it("normalizes native delay failures", async () => {
    const transport = new ContractTransport();
    const device = await probedDevice(transport);
    transport.failDelay = true;
    await expect(device.reset()).rejects.toBeInstanceOf(BusError);
  });
});
